File: 25. Kfolds.R
## K folds ##

# libraries ---------------------------------------------------------------

library(tidyverse)
library(caret)
library(kernlab)


# read data ---------------------------------------------------------------

datos <- readr::read_csv("data/Social_Network_Ads.csv")

# escalado de variables ---------------------------------------------------

datos <- datos %>% 
  dplyr::mutate(across(c(3,4), scale)) %>% 
  dplyr::mutate(across(c(3,4), as.numeric)) %>% 
  dplyr::select(3:5)

# kernel pca --------------------------------------------------------------

k_pca <- kernlab::kpca(~., datos[,-3], kernel = "rbfdot", 
                       features = 2)


datos_k_pca <- data.frame(k_pca@rotated, datos[, 3])



# validación cruzada ------------------------------------------------------

set.seed(123)

folds <- caret::createFolds(datos_k_pca$Purchased)

precisiones <- folds %>% 
  purrr::map_dbl(function(x){
    
    entrenamiento <- datos_k_pca[-x, ]
    validacion <- datos_k_pca[x, ]
    
    modelo <- entrenamiento %>% 
      glm(Purchased ~ ., data = ., family = "binomial")
    
    prediccion <- dplyr::if_else(predict(modelo, 
                                         validacion, 
                                         type = "response") > 0.5, 
                                 1, 0)
    
    cm <- table(validacion$Purchased, prediccion)
    
    sum(diag(cm))/sum(cm)
    
  })


mean(precisiones); sd(precisiones)


# ajustar clasificacion ---------------------------------------------------

clasificador <- entrenamiento %>% 
  glm(Purchased ~ ., data = ., family = "binomial")


# predicción de resultados ------------------------------------------------

y_pred <- dplyr::if_else(predict(clasificador, 
                                 validacion, type = "response") > 0.5, 
                         1, 0)

# matriz confusion --------------------------------------------------------

table(validacion$Purchased, y_pred)


# Visualizacion -----------------------------------------------------------

X1 <- seq(min(validacion$X1) - 1,
          max(validacion$X1) + 1, by = 0.05)

X2 <- seq(min(validacion$X2) - 1,
          max(validacion$X2) + 1, by = 0.05)

tidyr::expand_grid(X1, X2) %>% 
  predict(clasificador, newdata = . , type = "response") %>% 
  tibble::tibble(tidyr::expand_grid(X1, X2), Purchased = .) %>% 
  dplyr::mutate(Purchased = if_else(Purchased > 0.5, 1, 0)) %>% 
  ggplot2::ggplot(aes(X1, X2)) + 
  geom_tile(aes(fill = factor(Purchased)), alpha = 0.25) +
  geom_point(data = validacion, 
             aes(X1, X2, color = factor(Purchased))) +
  labs(title = "Clasificador: Logístico",
       subtitle = "(Conjunto de test)",
       x = "X1",
       y = "X2",
       fill = NULL,
       color = NULL) +
  scale_fill_brewer(palette = "Dark2") +
  scale_color_brewer(palette = "Dark2") +
  theme_classic()

File: 25. Kfolds.py
# -*- coding: utf-8 -*-
"""
Created on Fri Dec 30 10:25:31 2022

@author: Admin
"""

## K-Folds ##

# Regresion logística

# =============================================================================
# Libraries
# =============================================================================

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import sklearn.preprocessing as sklp
import sklearn.model_selection as sklm
import sklearn.linear_model as skll
import sklearn.metrics as sklmetric

# =============================================================================
# Read data
# =============================================================================

social = pd.read_csv("data\Social_Network_Ads.csv")

x = social.iloc[:, [2,3]].values
y = social.iloc[:, 4].values

# =============================================================================
# Escalado de variables
# =============================================================================

sc_x = sklp.StandardScaler()
x = sc_x.fit_transform(x)

# =============================================================================
# Reducir la dimensión con ACP
# =============================================================================

from sklearn.decomposition import KernelPCA

k_pca = KernelPCA(n_components = 2, kernel = "rbf")

x = k_pca.fit_transform(x)

# =============================================================================
# División de entrenamiento y test
# =============================================================================

x_train, x_test, y_train, y_test = sklm.train_test_split(x, y, test_size= 0.25,
                                                         random_state = 0)

# =============================================================================
# Ajustar modelo de regresion
# =============================================================================

clasificador = skll.LogisticRegression(random_state = 0)

clasificador.fit(x_train, y_train)

# =============================================================================
# Validación cruzada
# =============================================================================

from sklearn.model_selection import cross_val_score as CV

precisiones =  CV(estimator = clasificador, X = x_train, y = y_train, cv = 10)

precisiones.mean()
precisiones.std()

# =============================================================================
# Predicción de resultados 
# =============================================================================

y_pred = clasificador.predict(x_test)

# =============================================================================
# Matriz de confusion
# =============================================================================

cm = sklmetric.confusion_matrix(y_test, y_pred)
cm
# =============================================================================
# Representación gráfica
# =============================================================================

from matplotlib.colors import ListedColormap

x_set, y_set = x_test, y_test
X1, X2 = np.meshgrid(np.arange(start = x_set[:, 0].min() - 1,
                               stop = x_set[:, 0].max() + 1, step = 0.01),
                     np.arange(start = x_set[:, 1].min() - 1, 
                               stop = x_set[:, 1].max() + 1, step = 0.01))
plt.contourf(X1, X2, clasificador.predict(np.array([X1.ravel(), X2.ravel()]).T).reshape(X1.shape),
             alpha = 0.75, cmap = ListedColormap(('red', 'green')))
plt.xlim(X1.min(), X1.max())
plt.ylim(X2.min(), X2.max())
for i, j in enumerate(np.unique(y_set)):
    plt.scatter(x_set[y_set == j, 0], x_set[y_set == j, 1],
                c = ListedColormap(('red', 'green'))(i), label = j)
plt.title('Clasificador (Conjunto de Test)')
plt.xlabel('PCA1')
plt.ylabel('PCA2')
plt.legend()
plt.show()